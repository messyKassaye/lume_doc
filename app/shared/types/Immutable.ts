import Immutable from 'immutable';

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? Immutable.List<IImmutable<Elem>>
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
