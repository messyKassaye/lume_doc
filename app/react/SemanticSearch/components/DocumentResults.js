/* eslint-disable no-restricted-globals */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import Immutable from 'immutable';

import { Translate, t } from 'app/I18N';
import SnippetList from 'app/Documents/components/SnippetList';
import { selectSnippet } from 'app/Viewer/actions/uiActions';
import { Form } from 'react-redux-form';
import { NumericRangeSlide } from 'app/ReactReduxForms';
import { Icon } from 'app/Layout/Icon';
import { TemplateLabel, DocumentLanguage } from 'app/Layout';

const findResultsAboveThreshold = (results, threshold) => {
  const boundingIndex = results.findIndex(({ score }) => score < threshold);
  return boundingIndex >= 0 ? results.slice(0, boundingIndex) : results;
};

const getSnippetsFromResults = (results, template) => {
  const snippets = results.reduce(
    (_memo, item) => {
      const memo = _memo;
      const text = `${item.text} (${(item.score * 100).toFixed(2)}%)`;
      memo.count += 1;
      if (isNaN(Number(item.page))) {
        const field = template
          .get('properties')
          .find(p => p.get('name') === item.page)
          .get('label');
        memo.metadata[field] = (memo.metadata[field] || []).concat([text]);
        return memo;
      }
      memo.fullText.push({
        page: Number(item.page),
        text,
      });
      return memo;
    },
    { count: 0, metadata: {}, fullText: [] }
  );

  snippets.metadata = Object.keys(snippets.metadata).map(field => ({
    field,
    texts: snippets.metadata[field],
  }));
  return snippets;
};

export class DocumentResults extends Component {
  renderSnippetsList(doc, snippets, documentViewUrl) {
    return (
      <SnippetList
        doc={Immutable.fromJS(doc)}
        documentViewUrl={documentViewUrl}
        snippets={snippets}
        searchTerm=""
        selectSnippet={(...args) => {
          this.props.selectSnippet(...args);
        }}
      />
    );
  }

  renderFilter() {
    const { threshold } = this.props;
    return (
      <dl className="metadata-type-text">
        <dt className="item-header">
          <Translate>Threshold</Translate> {(threshold * 100).toFixed(2)} %
        </dt>
        <dd>
          <Form model="semanticSearch.resultsFilters">
            <NumericRangeSlide
              model=".threshold"
              min={0.3}
              max={1}
              step={0.01}
              delay={200}
              minLabel={t('System', 'Exploration')}
              maxLabel={t('System', 'Precision')}
            />
          </Form>
        </dd>
      </dl>
    );
  }

  render() {
    const { doc, threshold, template } = this.props;

    if (!doc.semanticSearch) {
      return false;
    }
    const filteredResults = findResultsAboveThreshold(doc.semanticSearch.results, threshold);
    const snippets = Immutable.fromJS(getSnippetsFromResults(filteredResults, template));
    const documentViewUrl = doc.file ? `/document/${doc.sharedId}` : `/entity/${doc.sharedId}`;

    return (
      <>
        <div className="view">
          <div className="item-info">
            <div>
              <Icon className="item-icon item-icon-center" data={doc.icon} />
              <h1 className="item-name">
                {doc.title}
                <DocumentLanguage doc={Immutable.fromJS(doc)} />
              </h1>
            </div>
            <TemplateLabel template={doc.template} />
          </div>
          {this.renderFilter()}
          <dl className="metadata-type-numeric">
            <dt>
              <Translate>Number of sentences above threshold</Translate>
            </dt>
            <dd>{filteredResults.length}</dd>
          </dl>
          <dl className="metadata-type-numeric">
            <dt>
              <Translate>% of sentences above threshold</Translate>
            </dt>
            <dd>
              {((filteredResults.length / doc.semanticSearch.totalResults) * 100).toFixed(2)}%
            </dd>
          </dl>
        </div>
        {this.renderSnippetsList(doc, snippets, documentViewUrl)}
      </>
    );
  }
}

DocumentResults.defaultProps = {
  template: undefined,
};

DocumentResults.propTypes = {
  doc: PropTypes.shape({
    title: PropTypes.string,
    template: PropTypes.string,
    sharedId: PropTypes.string,
    file: PropTypes.object,
    icon: PropTypes.object,
    semanticSearch: PropTypes.shape({
      totalResults: PropTypes.number,
      results: PropTypes.array,
    }),
  }).isRequired,
  threshold: PropTypes.number.isRequired,
  selectSnippet: PropTypes.func.isRequired,
  template: PropTypes.instanceOf(Immutable.Map),
};

const mapStateToProps = ({ semanticSearch, templates }) => ({
  threshold: semanticSearch.resultsFilters.threshold,
  template: templates.find(
    tpl => tpl.get('_id') === semanticSearch.selectedDocument.get('template')
  ),
});

function mapDispatchToProps(dispatch) {
  return bindActionCreators({ selectSnippet }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(DocumentResults);
