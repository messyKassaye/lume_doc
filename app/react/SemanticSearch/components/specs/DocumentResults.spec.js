import React from 'react';
import Immutable from 'immutable';
import { shallow } from 'enzyme';
import { DocumentResults } from '../DocumentResults';

describe('DocumentResults', () => {
  let props;
  let component;
  beforeEach(() => {
    props = {
      doc: {
        semanticSearch: {
          results: [
            { score: 0.6, text: 'bio two', page: 'bio' },
            { score: 0.5, text: 'page one', page: '1' },
            { score: 0.2, text: 'page two', page: '2' },
            { score: 0.1, text: 'bio one', page: 'bio' },
          ],
          relevantRate: 0.5,
          numRelevant: 2,
          totalResults: 4,
        },
        avgScore: 0.4,
      },
      threshold: 0.3,
      changeTreshHold: jasmine.createSpy('changeTreshhold'),
      selectTab: jasmine.createSpy('selectTab'),
      selectSnippet: jasmine.createSpy('selectSnippet'),
      template: Immutable.fromJS({
        commonProperties: [],
        properties: [{ name: 'bio', label: 'Biography' }],
      }),
    };

    component = shallow(<DocumentResults {...props} />);
  });

  describe('render', () => {
    it('should render results summary and snippets above threshold in SnippetsList', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
