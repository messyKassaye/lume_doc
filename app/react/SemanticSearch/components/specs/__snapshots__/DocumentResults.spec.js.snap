// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentResults render should render results summary and snippets above threshold in SnippetsList 1`] = `
<Fragment>
  <div
    className="view"
  >
    <div
      className="item-info"
    >
      <div>
        <Icon
          className="item-icon item-icon-center"
        />
        <h1
          className="item-name"
        >
          <Connect(DocumentLanguage)
            doc={
              Immutable.Map {
                "semanticSearch": Immutable.Map {
                  "results": Immutable.List [
                    Immutable.Map {
                      "score": 0.6,
                      "text": "bio two",
                      "page": "bio",
                    },
                    Immutable.Map {
                      "score": 0.5,
                      "text": "page one",
                      "page": "1",
                    },
                    Immutable.Map {
                      "score": 0.2,
                      "text": "page two",
                      "page": "2",
                    },
                    Immutable.Map {
                      "score": 0.1,
                      "text": "bio one",
                      "page": "bio",
                    },
                  ],
                  "relevantRate": 0.5,
                  "numRelevant": 2,
                  "totalResults": 4,
                },
                "avgScore": 0.4,
              }
            }
          />
        </h1>
      </div>
      <Connect(TemplateLabel) />
    </div>
    <dl
      className="metadata-type-text"
    >
      <dt
        className="item-header"
      >
        <Connect(Translate)>
          Threshold
        </Connect(Translate)>
         
        30.00
         %
      </dt>
      <dd>
        <Connect(Form)
          model="semanticSearch.resultsFilters"
        >
          <NumericRangeSlide
            delay={200}
            max={1}
            maxLabel="Precision"
            min={0.3}
            minLabel="Exploration"
            model=".threshold"
            step={0.01}
          />
        </Connect(Form)>
      </dd>
    </dl>
    <dl
      className="metadata-type-numeric"
    >
      <dt>
        <Connect(Translate)>
          Number of sentences above threshold
        </Connect(Translate)>
      </dt>
      <dd>
        2
      </dd>
    </dl>
    <dl
      className="metadata-type-numeric"
    >
      <dt>
        <Connect(Translate)>
          % of sentences above threshold
        </Connect(Translate)>
      </dt>
      <dd>
        50.00
        %
      </dd>
    </dl>
  </div>
  <Connect(SnippetList)
    doc={
      Immutable.Map {
        "semanticSearch": Immutable.Map {
          "results": Immutable.List [
            Immutable.Map {
              "score": 0.6,
              "text": "bio two",
              "page": "bio",
            },
            Immutable.Map {
              "score": 0.5,
              "text": "page one",
              "page": "1",
            },
            Immutable.Map {
              "score": 0.2,
              "text": "page two",
              "page": "2",
            },
            Immutable.Map {
              "score": 0.1,
              "text": "bio one",
              "page": "bio",
            },
          ],
          "relevantRate": 0.5,
          "numRelevant": 2,
          "totalResults": 4,
        },
        "avgScore": 0.4,
      }
    }
    documentViewUrl="/entity/undefined"
    searchTerm=""
    selectSnippet={[Function]}
    snippets={
      Immutable.Map {
        "count": 2,
        "metadata": Immutable.List [
          Immutable.Map {
            "field": "Biography",
            "texts": Immutable.List [
              "bio two (60.00%)",
            ],
          },
        ],
        "fullText": Immutable.List [
          Immutable.Map {
            "page": 1,
            "text": "page one (50.00%)",
          },
        ],
      }
    }
  />
</Fragment>
`;
